Closure compiler fixes for gl

nazar-pc opened the pull request:
As noted in https://github.com/kripken/emscripten/commit/93479ecbd390aec4f8a3765fe04bcb365d0b31b2#commitcomment-25520093 GL tests were failing.

Closure Compiler found 2 actual bugs in `src/library_gl.js` (which should ideally be covered by tests now):
* incorrect duplicated key `0x0300` instead of correct `0x0303` (correctness results from context)
* incorrect undefined `ext` variable used instead of `extension` (correctness results from context)

Also multiple undeclared/redeclared variables were either fixed or silenced (where they seem to be correct as is).

With this patch `browser.test_cubegeom_glew` passes for me locally. I recommend to review PR with `?w=1` as there are some trailing whitespaces removed in edited files.

Also see discussion here: https://github.com/kripken/emscripten/commit/97a464a654fdadf5dfb8aa082b48516e6bf8d402#commitcomment-25520648

juj commented on src/closure-externs.js:
```diff
@@ -1093,17 +1096,76 @@ var FUNCTION_TABLE;
 var define = function (wrapper) {};
+
+/**
+ * GL functions (https://github.com/kripken/emscripten/commit/97a464a654fdadf5dfb8aa082b48516e6bf8d402#commitcomment-25520648)
```
These are rather for emulated fixed pipeline GL functions, instead of general GL functions, but probably distinguishing in this file won't matter.

juj commented on src/gl-matrix.js:
```diff
@@ -726,9 +722,7 @@ mat4.determinant = function (mat) {
- * @param {mat4} [dest] mat4 receiving inverse matrix. If not specified result is written to mat
- *
- * @param {mat4} dest is specified, mat otherwise, null if matrix cannot be inverted
+ * @param {mat4} [dest] mat4 receiving inverse matrix. If not specified result is written to mat, null if matrix cannot be inverted
```
What are these edits for? `gl-matrix.js` comes from upstream https://github.com/toji/gl-matrix/blob/master/dist/gl-matrix.js, I think we should not do prose changes. Or are these just to accommodate a diagnostic from updated Closure?

nazar-pc replied:
These are to reduce number of warnings by Closure Compiler. They are not fatal errors though, so I think it is possible to revert changes is needed. Also linked upstream file seems to diverge from this one and doesn't have issues I've fixed here (time to update it?).

juj commented:
lgtm, though there's an in-flight merge conflict with another PR.

kripken commented:
I ran the browsers tests, and I still see these fail due to closure:
````
FAIL: test_html5 (test_browser.browser)
FAIL: test_html5_mouse (test_browser.browser)
FAIL: test_html5_webgl_create_context (test_browser.browser)
FAIL: test_sdl2glshader (test_browser.browser)
FAIL: test_sdlglshader (test_browser.browser)
FAIL: test_uuid (test_browser.browser)
FAIL: test_webgl2 (test_browser.browser)
FAIL: test_webgl_with_closure (test_browser.browser)
````
(But I'm to land this and leave the rest as followups, either way.)

nazar-pc commented:
Rebased against `incoming` and replaced `addEventListener` -> `window.addEventListener` change with new extern for Closure Compiler, since https://github.com/kripken/emscripten/commit/9193ca57852070532d029cbc2dc2a4c3f55f6511#diff-8e309db2313ebb40678cd543d93566f5R1060 also uses it this way and it is generally acceptable call for code that is common for regular browser environment and Web Worker.

kripken commented:
The `test_environ` [failure on the bots](https://s3.amazonaws.com/archive.travis-ci.org/jobs/300408669/log.txt?X-Amz-Expires=30&X-Amz-Date=20171111T022216Z&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRYRXRSVGNKPKO5A/20171111/us-east-1/s3/aws4_request&X-Amz-SignedHeaders=host&X-Amz-Signature=abefe032d92c2d457c1188e85f0f74cc5534b78d8c933b842319d463330b5020) looks like it is because of @juj's PR that changes the default language from C to C.UTF8.

nazar-pc commented:
I've fixed all of the mentioned tests here except `test_uuid`, as it is completely unrelated to GL, fix for it is in #5768.

I've also found one more bug in `src/library_gl.js` that was introduced here: https://github.com/kripken/emscripten/commit/de80d095559dc87857c94651b3ba4ca8aa1406fe#diff-78a51faf7f8c4f62113aae3d5d4ae8eeR1323
Fix for it seems correct from context, but test case for the fix would be nice to have, since it is not covered by any test yet.

juj commented:
> I've also found one more bug in src/library_gl.js that was introduced here: de80d09#diff-78a51faf7f8c4f62113aae3d5d4ae8eeR1323
> Fix for it seems correct from context, but test case for the fix would be nice to have, since it is not covered by any test yet.

Pushed fix to https://github.com/kripken/emscripten/commit/47fd4a1a5f53c58fc0651f9fd3a3f2e353b76a35. It would be nice to have test cases for things like this, but they are not trivial to develop. There is effort at Google going towards porting existing OpenGL ES test suite to run through Emscripten for full coverage.

juj commented:
> The test_environ failure on the bots looks like it is because of @juj's PR that changes the default language from C to C.UTF8.

Thanks, pushed https://github.com/kripken/emscripten/commit/5a7b1ec6ea049fd3affd2b7f8ec0a4ec9b9e8762.

nazar-pc commented:
Rebased once again, let's see what CI will say now.

nazar-pc commented:
OK, now only 2 tests are failing according to CI and I don't think they are related to this PR (both also fail on `incoming` as of recently):
```
======================================================================
ERROR: test_simd3 (test_core.default)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/root/emscripten/tests/test_core.py", line 18, in decorated
    f(self)
  File "/root/emscripten/tests/test_core.py", line 5182, in test_simd3
    self.do_run_from_file(src, output)
  File "/root/emscripten/tests/runner.py", line 643, in do_run_from_file
    assert_returncode, assert_identical)
  File "/root/emscripten/tests/runner.py", line 683, in do_run
    raise e
Exception: Expected to find 'DONE' in 'warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
exception thrown: Assertion failed: ar2[0] == 256.0, at: /tmp/emscripten_test_default_hlQjHY/src.cpp,139,testConversions at Error
    at jsStackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2524:13)
    at stackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2541:12)
    at ___assert_fail (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:3100:210)
    at __Z15testConversionsv (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:4233:3)
    at _main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:6930:2)
    at Object.asm._main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13449:21)
    at Object.callMain (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13722:30)
    at doRun (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13786:60)
    at run (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13800:5)
    at Object.<anonymous> (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13878:1)

/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:1485
      throw ex;
      ^
Assertion failed: ar2[0] == 256.0, at: /tmp/emscripten_test_default_hlQjHY/src.cpp,139,testConversions at Error
    at jsStackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2524:13)
    at stackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2541:12)
    at ___assert_fail (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:3100:210)
    at __Z15testConversionsv (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:4233:3)
    at _main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:6930:2)
    at Object.asm._main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13449:21)
    at Object.callMain (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13722:30)
    at doRun (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13786:60)
    at run (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13800:5)
    at Object.<anonymous> (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13878:1)
', diff:

--- expected
+++ actual
@@ -1 +1,28 @@
-DONE
+warning: build with  -s DEMANGLE_SUPPORT=1  to link in libcxxabi demangling
+exception thrown: Assertion failed: ar2[0] == 256.0, at: /tmp/emscripten_test_default_hlQjHY/src.cpp,139,testConversions at Error
+    at jsStackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2524:13)
+    at stackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2541:12)
+    at ___assert_fail (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:3100:210)
+    at __Z15testConversionsv (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:4233:3)
+    at _main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:6930:2)
+    at Object.asm._main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13449:21)
+    at Object.callMain (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13722:30)
+    at doRun (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13786:60)
+    at run (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13800:5)
+    at Object.<anonymous> (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13878:1)
+
+/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:1485
+      throw ex;
+      ^
+Assertion failed: ar2[0] == 256.0, at: /tmp/emscripten_test_default_hlQjHY/src.cpp,139,testConversions at Error
+    at jsStackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2524:13)
+    at stackTrace (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:2541:12)
+    at ___assert_fail (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:3100:210)
+    at __Z15testConversionsv (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:4233:3)
+    at _main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:6930:2)
+    at Object.asm._main (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13449:21)
+    at Object.callMain (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13722:30)
+    at doRun (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13786:60)
+    at run (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13800:5)
+    at Object.<anonymous> (/tmp/emscripten_test_default_hlQjHY/src.cpp.o.js:13878:1)
+



======================================================================
ERROR: test_simd_sitofp (test_core.default)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/root/emscripten/tests/test_core.py", line 18, in decorated
    f(self)
  File "/root/emscripten/tests/test_core.py", line 5314, in test_simd_sitofp
    self.do_run_in_out_file_test('tests', 'core', 'test_simd_sitofp')
  File "/root/emscripten/tests/test_core.py", line 81, in do_run_in_out_file_test
    self.do_run_from_file(src, output, **kwargs)
  File "/root/emscripten/tests/runner.py", line 643, in do_run_from_file
    assert_returncode, assert_identical)
  File "/root/emscripten/tests/runner.py", line 683, in do_run
    raise e
Exception: Expected to find '255.000000' in '0.000000
', diff:

--- expected
+++ actual
@@ -1 +1,2 @@
-255.000000
+0.000000
+
```

kripken commented:
Those simd errors can be ignored, they are because of recent changes and require a new fastcomp build for the emsdk we use on travis.

kripken commented:
Btw, instead of rebasing, you can just merge. That's usually easier and also means less to review each time. We squash it all before the final merge anyhow.

kripken commented:
Ok, let's merge this in, and later see how it looks after merging it into the browser-travis PR #5770
